File: Kotlin/app/src/main/kotlin/ca/jwolf/cubesolver/utils/ProgramUtils.kt
package ca.jwolf.cubesolver.utils

import ca.jwolf.cubesolver.benchmarking.BenchmarkSettings
import ca.jwolf.cubesolver.configs.PuzzleConfig
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.MapperFeature
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.databind.json.JsonMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.fasterxml.jackson.module.kotlin.registerKotlinModule

import java.nio.file.Paths
import kotlin.system.exitProcess

class ProgramUtils {

    companion object {
        private const val LINE_LENGTH = 100
        private var indent = 0
        val jsonMapper = JsonMapper.builder()
            .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .enable(SerializationFeature.INDENT_OUTPUT)
            .build().registerKotlinModule()

        var ConfigurationDirectory = ""
            private set

        var OutputDirectory = ""

        var BenchmarkConfigurationDirectory = ""
            private set

        fun indent() {
            indent += 2
        }

        fun unindent() {
            indent -= 2
        }

        fun SetupConfigurationDirectory(directory: String) {
            ConfigurationDirectory = directory
            BenchmarkConfigurationDirectory = Paths.get(ConfigurationDirectory, "Benchmarks").toString()
        }

        fun writeLine(text: String ="", actuallyWrite: Boolean = true)
        {
            if (!actuallyWrite) {
                return
            }

            val prefix = " ".repeat(indent)
            println(prefix + text)
        }

        fun reWrite(text: String="", actuallyWrite: Boolean = true)
        {
            if (!actuallyWrite) {
                return
            }

            val prefix = " ".repeat(indent)
            print("\r$prefix$text")
        }

        fun writeFilledLine(fillCharacter: Char, text: String="", actuallyWrite: Boolean = true)
        {
            if (!actuallyWrite) {
                return
            }


            val targetLineLength = LINE_LENGTH - indent
            val textLength = text.length

            if (textLength == 0)
            {
                writeLine(fillCharacter.toString().repeat(targetLineLength))
            }
            else
            {
                val middleLineLength = (targetLineLength - textLength - 2)/2
                val filler = fillCharacter.toString().repeat(middleLineLength)
                writeLine("$filler $text $filler")
            }

        }

        fun getUserInputString(prompt: String, defaultValue: String="") : String
        {
            val inputPrompt = ">> $prompt "
            val prefix = " ".repeat(indent)
            var defaultedValue = ""
            if (defaultValue.isNotBlank()) defaultedValue = "($defaultValue) "
            print("$prefix$inputPrompt$defaultedValue")

            var input = readLine()
            input = input!!.replace(inputPrompt, "").trim()
            if (input.isBlank()) return defaultValue
            return input
        }

        fun getUserInputInt(prompt: String, defaultValue: Int=-1) : Int
        {
            val input = getUserInputString(prompt, defaultValue.toString())
            return input.toIntOrNull() ?: defaultValue
        }

        fun getUserSelection(context: String, options: List<String>) : Int
        {
            writeLine(":::: $context ::::")
            var optionIndex = 0
            for (option in options)
            {
                optionIndex++
                writeLine("$optionIndex: $option")
            }

            val lastOptionIndex = optionIndex
            optionIndex++
            writeLine("$optionIndex: Exit")

            val selected = getUserInputInt("Select an option: ")
            if (0 >= selected || selected > lastOptionIndex)
            {
                exitProcess(0)
            }


            return (selected - 1)
        }

        fun getConfigFromUser(): PuzzleConfig {
            val fileList = Paths.get(ConfigurationDirectory).toFile().listFiles()!!.map { it.name }.filter { it.endsWith(".json") }
            val configFile = getUserSelection("Select a config file to load:", fileList)

            val config =  jsonMapper.readValue<PuzzleConfig>(Paths.get(ConfigurationDirectory, fileList[configFile]).toFile().readBytes().decodeToString())
            config.initializePieces()
            return config
        }

        fun getBenchmarkSettingsFromUser(): BenchmarkSettings {
            val fileList = Paths.get(BenchmarkConfigurationDirectory).toFile().listFiles()!!.map { it.name }.filter { it.endsWith(".json") }
            val configFile = getUserSelection("Select a benchmark file to load:", fileList)

            val config =  jsonMapper.readValue<BenchmarkSettings>(Paths.get(BenchmarkConfigurationDirectory, fileList[configFile]).toFile().readBytes().decodeToString())
            return config
        }

        //extension methods
        fun Long.formatWithCommas(): String = String.format("%,d", this)
        fun Int.formatWithCommas(): String = String.format("%,d", this)

    }

}

File: CSharp/CubeSolver/Utils/ProgramUtils.cs
using System.Text.Json;
using NetCubeSolver;
using NetCubeSolver.Benchmarking;

namespace CubeSolver.Utils;

public class ProgramUtils
{
    static int LINE_LENGTH = 100;

    private static int _indent = 0;
    
    public static string WorkingDirectory { get; private set; }
    public static string OutputDirectory;
    public static string BenchmarkDirectory { get; private set; }

    public static void SetupWorkingDirectory(string workingDirectory)
    {
        WorkingDirectory = workingDirectory;
        BenchmarkDirectory = Path.Combine(WorkingDirectory, "Benchmarks");
    }

    public static void Write(string text="")
    {
        var prefix = string.Concat(Enumerable.Repeat(" ", _indent));
        Console.WriteLine($"{prefix}{text}");
    }

    public static void ReWrite(string text = "")
    {
        var prefix = string.Concat(Enumerable.Repeat(" ", _indent));
        Console.Write($"\r{prefix}{text}");
    }

    public static void WriteFilledLine(char fillCharacter, string line="")
    {
        var targetLineLength = LINE_LENGTH - _indent;
        var lineLength = line.Length;

        if (lineLength == 0)
        {
            Write(string.Concat(Enumerable.Repeat(fillCharacter, targetLineLength)));
        }
        else
        {
            var middleLineLength = (targetLineLength - lineLength - 2) / 2;
            var filler = string.Concat(Enumerable.Repeat(fillCharacter, middleLineLength));
            Write($"{filler} {line} {filler}");
        }
    }

    public static string GetUserInput(string prompt, string defaultValue = "")
    {
        var inputPrompt = $">> {prompt} ";
        var prefix = string.Concat(Enumerable.Repeat(" ", _indent));
        var defaultedValue = "";
        if (!string.IsNullOrWhiteSpace(defaultValue))
        {
            defaultedValue = $"({defaultValue}) ";
        }
        Console.Write($"{prefix}{inputPrompt}{defaultedValue}");
        var input = Console.ReadLine();
        Write();
        
        return input?.Replace(inputPrompt, "").Trim();
    }

    public static int GetUserInputInt(string prompt, int defaultValue = 1)
    {
        var input = GetUserInput(prompt, defaultValue.ToString());
        var selection = defaultValue;
        try
        {
            if (string.IsNullOrWhiteSpace(input) || string.IsNullOrEmpty(input))
            {
                return defaultValue;
            }
            selection = int.Parse(input ?? defaultValue.ToString());
        }
        catch (Exception e)
        {
            Environment.Exit(0);
        }

        return selection ;
    }

    public static int GetUserSelection(string context, string[] options)
    {
        Write($":::: {context} ::::");
        var optionIndex = 0;
        foreach (var option in options)
        {
            Write($" {optionIndex + 1}: {option}");
            optionIndex++;
        }
        
        Write($" {optionIndex + 1}: Exit ");
        
        var input = GetUserInput("Select an option: ");
        var selection = 1;
        try
        {
            selection = int.Parse(input ?? "-1");
            if (selection == optionIndex + 1)
            {
                Environment.Exit(0);
            }
        }
        catch (Exception e)
        {
            Environment.Exit(0);
        }

        return selection - 1;
    }

    public static void Indent() =>  _indent+=2;

    public static void Unindent()
    {
        _indent -= 2;
        _indent = Math.Max(0, _indent);
    }

    public static Config GetConfigFromUser()
    {
        DirectoryInfo dir = new DirectoryInfo(WorkingDirectory);
        var configs = dir.GetFiles().Where(f => f.Extension == ".json").ToList();
        var configNames = configs.Select(x => x.Name).ToArray();
        var configFile = GetUserSelection("Select a config file to load:", configNames);
        
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
        
        Config config = JsonSerializer.Deserialize<Config>(File.ReadAllText(configs[configFile].FullName), options);
        config.InitializePieces();
        return config;
        
    }
    
    public static BenchmarkSettings GetBenchmarkSettingsFromUser()
    {
        DirectoryInfo dir = new DirectoryInfo(BenchmarkDirectory);
        var configs = dir.GetFiles().Where(f => f.Extension == ".json").ToList();
        var configNames = configs.Select(x => x.Name).ToArray();
        var configFile = GetUserSelection("Select a benchmark file to load:", configNames);
        
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
        
        BenchmarkSettings settings = JsonSerializer.Deserialize<BenchmarkSettings>(File.ReadAllText(configs[configFile].FullName), options);
        return settings;
        
    }
}